src/main/scala/viper/gobra/translator/encodings/slices/SliceEncoding.scala

```diff
@@ -0,0 +1,461 @@
+    case ctx.Slice(t) / m => m match {
+      case Exclusive => ctx.option.typ(ctx.slice.typ(ctx.typeEncoding.typ(ctx)(t)))
+      case Shared => vpr.Ref
+    }
+  }
+
+  override def expr(ctx : Context) : in.Expr ==> CodeWriter[vpr.Exp] = {
+    def goE(x: in.Expr): CodeWriter[vpr.Exp] = ctx.expr.translate(x)(ctx)
+
+    default(super.expr(ctx)) {
+      case (exp : in.DfltVal) :: ctx.Slice(t) / m => {
+        val (pos, info, errT) = exp.vprMeta
+        m match {
+          case Exclusive => unit(nilSlice(t)(ctx)(pos, info, errT))
+          case Shared => unit(vpr.NullLit()(pos, info, errT))
+        }
+      }
+
+      case (exp : in.NilLit) :: ctx.Slice(t) / Exclusive => {
+        val (pos, info, errT) = exp.vprMeta
+        unit(nilSlice(t)(ctx)(pos, info, errT))
+      }
+
+      case in.Length(exp :: ctx.Slice(typ)) => for {
+        expT <- goE(exp)
+        typT = ctx.typeEncoding.typ(ctx)(typ)
+        (pos, info, errT) = exp.vprMeta
+      } yield vpr.CondExp(
+        vpr.EqCmp(expT, nilSlice(typT)(ctx)(pos, info, errT))(pos, info, errT),
+        vpr.IntLit(0)(pos, info, errT),
+        ctx.slice.len(ctx.option.get(expT, ctx.slice.typ(typT))(pos, info, errT))(pos, info, errT)
+      )(pos, info, errT)
+
+      case in.Capacity(exp :: ctx.Slice(typ)) => for {
+        expT <- goE(exp)
+        typT = ctx.typeEncoding.typ(ctx)(typ)
+        (pos, info, errT) = exp.vprMeta
+      } yield vpr.CondExp(
+        vpr.EqCmp(expT, nilSlice(typT)(ctx)(pos, info, errT))(pos, info, errT),
+        vpr.IntLit(0)(pos, info, errT),
+        ctx.slice.cap(ctx.option.get(expT, ctx.slice.typ(typT))(pos, info, errT))(pos, info, errT)
+      )(pos, info, errT)
+
+      case exp @ in.Slice((base : in.Location) :: ctx.Array(_, _) / Shared, low, high, max) => for {
+        baseT <- ctx.typeEncoding.reference(ctx)(base)
+        lowT <- goE(low)
+        highT <- goE(high)
+        maxOptT <- option(max match {
+          case Some(e) => Some(goE(e))
+          case None => None
+        })
+      } yield {
+        val (pos, info, errT) = exp.vprMeta
+        val unboxedBaseT = arrayEmb.unbox(baseT, base.typ.asInstanceOf[in.ArrayT])(base)(ctx)
+
+        ctx.option.some(maxOptT match {
+          case None => sliceFromArray(vpr.Ref, unboxedBaseT, lowT, highT)(ctx)(pos, info, errT)
+          case Some(maxT) => fullSliceFromArray(vpr.Ref, unboxedBaseT, lowT, highT, maxT)(ctx)(pos, info, errT)
+        })(pos, info, errT)
+      }
+
+      case exp @ in.Slice((base : in.Expr) :: ctx.Slice(typ), low, high, max) => for {
+        baseT <- goE(base)
+        lowT <- goE(low)
+        highT <- goE(high)
+        maxOptT <- option(max match {
+          case Some(e) => Some(goE(e))
+          case None => None
+        })
+      } yield {
+        val (pos, info, errT) = exp.vprMeta
+        val typT = ctx.typeEncoding.typ(ctx)(typ)
+        val baseOptT = ctx.option.get(baseT, ctx.slice.typ(typT))(pos, info, errT)
```

Comment:
What happens when the slice is nil?

Reply:
There are some problems here I see...

This for example currently doesn't verify but should:

```go
requires s == nil
func foo(s []int) {
  assert s[:] == nil
}
```
And this would actually crash Gobra:
```go
requires s == nil
func foo(s []int) {
  assert s[1:] == nil
}
```

I'll have a look at that, thanks!

Reply:
The first case is very interesting. If you are reusing my nil for shared arryas, then we should change that nil has a size of zero (currently it is a size of 1). Indeed, calling len and cap on a nil slice does return 0.

Reply:
Why? One could define the nil slice to be equivalent to `nilArray[0:0:0]`, even when `len(nilArray) = 1`.

Reply:
Aah wait, I think I see why. Then the nil array could potentially be sliced to have one element.

Reply:
You could prevent this with an additional check in the encoding of the slice expression. 
I am not entirely sure myself if nil should have size 1 or 0. 
When you say that nil has size 0, then it hard to distinguish a zero sized shared array and nil.
I think that this is also the reason why I picked size 1 for nil.

Reply:
With the current way default arrays are encoded, I don't think they can be of length 0, can they?

If the encoding would be
```viper
function arrayDefault(): ShArray[Ref]
  ensures (ShArraylen(result): Int) == 0
  ensures (forall i: Int :: { ShArrayloc(result, i) } ShArrayloc(result, i) == null)
```
Then the `nil` slice might no longer be unique.

Reply:
I do not entirely understand what you mean. Why would nil not be unique anymore. Because of arrays of size 0? I think that this depends on the equality relation. Since we do not have partial functions, two arrays are different even if their size is 0, but their elements are different. So as long as 0-sized arrays do not have array locations that are nil (which every non-nil array must satisfy), we should be fine.

Reply:
What I meant is that, for any two arrays `a1` and `a2`, both of length 0 and both having `null` at every location, one cannot prove that `a1 == a2`, right? And this can be proven for any length greater than 0.

Reply:
Perhaps the way the equality relation is defined could resolve that. But I don't yet see how.

Could we not change the injectivity axiom for arrays to be the following (i.e., dropping the bounds)?
```viper
axiom {
  forall a : Array[T], i : Int :: { loc(a, i) } fst(loc(a, i)) == a && snd(loc(a, i)) == i
}
```

Reply:
The purpose of the bounds is to prevent an unsoundness in combination with nil.
Regarding nil, the property that we want is that if an array location is nil, then the array is nil.
This is equivalent to that 1) the nil array has nil locations for every index and 2) only the nil array has nil entries.
However, this violates injectivity of loc, since a nil location must identify the corresponding array and the corresponding index. 
I saw two solutions to solve this: 1) make injectivity hold for all arrays that are not nil or 2) make injectivity hold only on valid indexes and give nil either size 1 or 0.